app/Http/Controllers/Admin/RoleController.php

```diff
@@ -5,82 +5,95 @@
-     *
-     * @return \Illuminate\Http\Response
-     */
     public function create()
     {
-        return view('admin.role.create');
+        return view('admin.role.form', [
+            'permissionGroups' => $this->permissionGroupRepository->with('permissions')->get(),
+        ]);
     }
 
-    /**
-     * Store a newly created resource in storage.
-     *
-     * @param  \Illuminate\Http\Request  $request
-     * @return \Illuminate\Http\Response
-     */
     public function store(RoleRequest $request)
     {
-        //
+        DB::beginTransaction();
+        try {
+            $role = $this->roleRepository->save($request->validated());
+            $role->permissions()->sync($request->input('permission_ids'));
+            DB::commit();
+            return redirect()->route('admin.role.show', $role->id);
+        } catch (Exception $roles) {
+            DB::rollBack();
+
+            throw new Exception($roles->getMessage('massage','error, please try again'));
+        }
+
     }
 
-    /**
-     * Display the specified resource.
-     *
-     * @param  int  $id
-     * @return \Illuminate\Http\Response
-     */
     public function show($id)
     {
-        //
+        if (! $role = $this->roleRepository->findById($id)) {
+            abort(404);
+        }
+
+        return view('admin.role.show', [
+            'role' => $role,
+        ]);
     }
 
-    /**
-     * Show the form for editing the specified resource.
-     *
-     * @param  int  $id
-     * @return \Illuminate\Http\Response
-     */
     public function edit($id)
     {
-        //
+        if (! $role = $this->roleRepository->findById($id)) {
+            abort(404);
+        }
+
+        return view('admin.role.form', [
+            'role' => $role,
+            'permissionGroups' => $this->permissionGroupRepository->getAll(),
+        ]);
     }
 
-    /**
-     * Update the specified resource in storage.
-     *
-     * @param  \Illuminate\Http\Request  $request
-     * @param  int  $id
-     * @return \Illuminate\Http\Response
-     */
-    public function update(Request $request, $id)
+    public function update(RoleRequest $request, $id)
     {
-        //
+        DB::beginTransaction();
+        try {
+            $role = $this->roleRepository->save($request->validated(), ['id' => $id]);
+            $role->permissions()->sync($request->input('permission_ids'));
+
+            DB::commit();
+            return redirect()->route('admin.role.index');
+        } catch (Exception $roles) {
+            DB::rollBack();
+
+            throw new Exception($roles->getMessage('massage','error, please try again'));
+        }
+
+
     }
 
-    /**
-     * Remove the specified resource from storage.
-     *
-     * @param  int  $id
-     * @return \Illuminate\Http\Response
-     */
     public function destroy($id)
     {
-        //
+        $this->roleRepository -> deleteById($id);
+        return redirect()->route('admin.role.index');
```

Comment:
`$this->roleRepository -> deleteById($id);`
=>
`$this->roleRepository->deleteById($id);`

Reply:
trước khi xoá role anh phải xoá relation trước nhé, tránh lỗi

Reply:
empty line before return